### ISSUE TITLE
CPM: Issues with mass_conv_core reporting

### ISSUE THREAD

**Issue by rjfarmer** [MESAHub/mesa#30]
From Anne:

It seems like that with the new CPM there are issues with reporting history items like mass_conv_core and conv_mx1_top are not reported properly and do not map to the actual convective/semiconvective boundary.

Investigating

**Comment by rjfarmer** [comment]
With the re-addition of the convective_bdy_weight controls is this fixed now or at least better?

**Comment by evbauer** [comment]
I think this is a somewhat different issue conceptually. I would want to check with @annethoul to be sure, but my recollection is that there was some inconsistency between what MESA reports as the core boundary and the location of material that gets mixed by semiconvection when CPM is on.

**Comment by annethoul** [comment]
Yes, Evan is correct. These issues are totally separate. convective_bdy_weight is something that actually interferes with the models, to evolve the convective boundary more smoothly from timestep to timestep. The problem described in this thread is that the models are good, and if you use profiles to figure out the convective (and "partially mixed" ) boundaries, you get the correct positions, while those boundaries  are incorrectly reported in the history files. For MESA 5 figures, Rich figured out the boundaries using the profiles. But apparently it was not straightforward to do. Still waiting for a volunteer to write a piece of code that finds those boundaries from the profiles data (from I assume, kinks in the abundance profiles?)

**Comment by evbauer** [comment]
@annethoul I think I recall you had some plots that pretty specifically isolate where the location should be based on the profiles vs where it is reported in the history files. Do you have those available to post here? Or better yet, could you generate a minimal example on the latest release?

**Comment by annethoul** [comment]
Here is one example (inlist based on the mass_conv_core test suite example, with better resolution). Looking at the history plot of mass_conv_core versus model number, one sees "peaks" or "dips" in the convective core mass. Those are due to tiny splittings of the core. CPM very quickly recovers from those small splittings, and by eye one can easily follow the evolution of the convective core mass, but in the history file, the value reported for the mass of the convective core is the first convective zone from the center. Note that those tiny splittings also mess up the profile of he4, and it is only when the core size increases and "swallows" the messed up region that we recover a decent profile for he4. The exact evolution of he4 is probably not totally correct due to those small splittings.

**Comment by annethoul** [comment]
history
<img width="889" alt="history" src="https://github.com/MESAHub/mesa/assets/56915840/c7d59eb6-d7bc-46cd-a41a-931aebac52a8">

**Comment by annethoul** [comment]
profile 1256
<img width="893" alt="profile1256" src="https://github.com/MESAHub/mesa/assets/56915840/9de9bb46-871b-4def-a6f4-16515ac4ff9b">

**Comment by annethoul** [comment]
profile 1257
<img width="868" alt="profile1257" src="https://github.com/MESAHub/mesa/assets/56915840/aacee977-ed18-49da-bac2-a6cf5ba11dde">

**Comment by annethoul** [comment]
profile 1258
<img width="858" alt="profile1258" src="https://github.com/MESAHub/mesa/assets/56915840/bf0996ad-f85d-43ae-a475-6e4c48ea451e">

**Comment by annethoul** [comment]
profile 1350
<img width="865" alt="profile1350" src="https://github.com/MESAHub/mesa/assets/56915840/8d44a9d9-e8ca-4d00-8070-76594ea7ff26">

**Comment by annethoul** [comment]
@wmwolf Note: all figures generated using Bill's all new and wonderful MESA Explorer tool!

**Comment by annethoul** [comment]
Also, kind of separate issue but linked to this one, once a "semiconvective" region (due to CPM) develops, MESA has no way to locate its boundary.

**Comment by fxt44** [comment]
> mass_conv_core test suite example

i don't see this test suite case in 15140 or 23.05.1. are you pulling it from another test suite case? or maybe i'm just being silly.

> Those are due to tiny splittings of the core.

one of your hand-drawn figs explains it pretty well:
![IMG_8071_1200](https://github.com/MESAHub/mesa/assets/7750797/93046592-101b-4dc9-b672-253b3bdc341f)



> Note that those tiny splittings also mess up the profile of he4, and it is only when the core size increases and "swallows" the messed up region that we recover a decent profile for he4.

with crazy small timesteps and high mass resolution, it is possible to avoid the splitting until x(he4) drops below ~1e-3, which is still an issue if one cares about the central o16/c12. its more fun further out in the star :)

**Comment by annethoul** [comment]
@fxt44 my fault, it is called conv_core_cpm ... told you I am getting old ;-(

**Comment by evbauer** [comment]
> Also, kind of separate issue but linked to this one, once a "semiconvective" region (due to CPM) develops, MESA has no way to locate its boundary. Again, easy to locate "by eye" looking at e.g. he4 profiles, but hard to code.

Oh interesting. This is what I thought the main point of this issue was. This part is strictly a reporting issue, in the sense that we know where the boundary ought to be intuitively, but still have some work to do to get MESA to report the right number.

The other issue is more of an actual physics issue that's having an impact on the evolution of the models, and probably should motivate some sort of improved resolution controls to prevent the splitting from happening.
